devicetypes/smartthings/fibaro-rgbw-controller.src/fibaro-rgbw-controller.groovy

```diff
@@ -26,562 +26,236 @@
-        value.gh = hex(value.green * value.level/100)
-        value.bh = hex(value.blue * value.level/100)
-        value.hex = "#${hex(value.red)}${hex(value.green)}${hex(value.blue)}"
-    }
-
-    if(value.hue) {
-        sendEvent(name: "hue", value: value.hue, displayed: false)
-    }
-    if(value.saturation) {
-        sendEvent(name: "saturation", value: value.saturation, displayed: false)
-    }
-    if(value.hex?.trim()) {
-        sendEvent(name: "color", value: value.hex, displayed: false)
-    }
-    if (value.level) {
-        sendEvent(name: "level", value: value.level)
-    }
-    if (value.switch?.trim()) {
-        sendEvent(name: "switch", value: value.switch)
-    }
-
-    sendRGB(value.rh, value.gh, value.bh)
-}
-
-def setLevel(level, rate = null) {
-	log.debug "setLevel($level)"
-
-	if (level == 0) { off() }
-	else if (device.latestValue("switch") == "off") { on() }
-
-    def colorHex = device.latestValue("color")
-    if (colorHex == null)
-		colorHex = "#FFFFFF"
-
-    def c = hexToRgb(colorHex)
-
-    def r = hex(c.r * (level/100))
-    def g = hex(c.g * (level/100))
-    def b = hex(c.b * (level/100))
-
-	sendEvent(name: "level", value: level)
-    sendEvent(name: "setLevel", value: level, displayed: false)
-	sendRGB(r, g, b)
-}
-
-
-def setWhiteLevel(value) {
-	log.debug "setWhiteLevel: ${value}"
-    def level = Math.min(value as Integer, 99)
-    level = 255 * level/99 as Integer
-    def channel = 0
+	def rgb
+	if (value.hex) {
+		rgb = value.hex.findAll(/[0-9a-fA-F]{2}/).collect { Integer.parseInt(it, 16) }
+	} else if (value.hue != null && value.saturation != null) {
+		rgb = huesatToRGB(value.hue, value.saturation)
+	}
 
-	if (device.latestValue("switch") == "off") { on() }
+	if (rgb) {
+		result << zwave.switchColorV3.switchColorSet(red: rgb[0], green: rgb[1], blue: rgb[2])
+	}
 
-    sendEvent(name: "whiteLevel", value: value)
-    sendWhite(channel, value)
-}
+	if (value.hex) sendEvent(name: "color", value: value.hex)
+	if (value.hue) sendEvent(name: "hue", value: value.hue)
+	if (value.saturation) sendEvent(name: "saturation", value: value.saturation)
 
-def sendWhite(channel, value) {
-	def whiteLevel = hex(value)
-    def cmd = [String.format("3305010${channel}${whiteLevel}%02X", 50)]
-    cmd
+	commands(result)
 }
 
-def sendRGB(redHex, greenHex, blueHex) {
-    def cmd = [String.format("33050302${redHex}03${greenHex}04${blueHex}%02X", 100),]
-    cmd
+def setLevel(level, duration=null) {
+	level = Math.max(Math.min(level, 99), 0)
+	commands([zwave.basicV1.basicSet(value: level), zwave.switchMultilevelV3.switchMultilevelGet()], 2000)
 }
 
+private getCOLOR_TEMP_MAX() { 100 }
+private getCOLOR_TEMP_MIN() { 0 }
+private getCOLOR_TEMP_DIFF() { COLOR_TEMP_MAX - COLOR_TEMP_MIN }
 
-def sendRGBW(redHex, greenHex, blueHex, whiteHex) {
-    def cmd = [String.format("33050400${whiteHex}02${redHex}03${greenHex}04${blueHex}%02X", 100),]
-    cmd
+def setColorTemperature(temp) {
```

Comment:
This also should support K instead of percentage.

Reply:
Device doesn't have a defined Kelvin range anyway, so anything I did would just be made up. To boot, default range for the color temp widget in fallback metadata on OneApp is 1-100 (listed as K). Idk. If you feel strongly about it I can change it back and generate metadata.

Reply:
I think it's important to be consistent with the other devices as much as we can. If this device doesn't support a specific color temperature range then let's go with 2700-6500 which is our general default. 

I'm not familiar with this device or how Z-Wave deals with color temperature. Does this device support both warmWhite and coldWhite components? Can they be "mixed" to achieve values between warm and cold? My guess is we can map 2700K to 100% warm white and 6500K to 100% cold white but I'm not sure how to get the inbetween values.

Reply:
Based on testing (and the previous DTH) only the `warmWhite` component is supported. I think the way the device interprets it is that 0% `warmWhite` is `coldWhite`.